sphinx/tutorial/rose/furthertopics/optional-configurations.rst

```diff
@@ -0,0 +1,181 @@
+``rose task-run`` is effectively a wrapper for ``rose app-run``.
+
+Have a look at the help for ``rose app-run``, by running
+``rose help app-run``. You can see that we can use ``--define`` to
+override things in the app at runtime.
+
+In the ``suite.rc`` file, modify the ``script`` for the ``order_icecream``
+task so that it reads:
+
+.. code-block:: cylc
+
+   script = rose task-run --define='[env]FLAVOUR=chocolate' --define='[env]TOPPING=fudge'
+
+This will override the ``FLAVOUR`` and ``TOPPING`` options in the ``env``
+section of our app. If you run the suite again, you should see that the
+output of the task has changed.
+
+If we had any more options to override, it might get a little unwieldy - it
+would be nicer to create an optional configuration containing these
+settings that we could apply on-demand, especially if they were used regularly.
+
+Create a subdirectory ``opt/`` within the app itself
+(``app/order_icecream/opt/``). This is where Rose will look for optional
+configurations. Individual optional configuration files (there can be
+more than one in the same directory) are named ``rose-app-NAME.conf``
+where ``NAME`` is some kind of useful id.
+
+
+New Optional Configurations
+---------------------------
+
+Create an optional configuration file called ``rose-app-chocofudge.conf``
+in your new ``opt/`` subdirectory, with the following content:
+
+.. code-block:: rose
+
+   [env]
+   FLAVOUR=chocolate
+   TOPPING=fudge
+
+We can now reference this through an option to ``rose task-run`` - alter
+the ``script`` of the ``order_icecream`` task (in the ``suite.rc`` file)
+to read:
+
+.. code-block:: cylc
+
+   script = rose task-run --opt-conf-key=chocofudge
+
+This will now override the ``order_icecream`` app's ``rose-app.conf``
+file with the contents of the ``rose-app-chocofudge.conf`` file at runtime.
+
+We can add as many ``--opt-conf-key`` options to ``rose task-run`` as we
+like - not just the one. This would apply each optional configuration in
+first-to-last order.
+
+Try running the suite - it should apply the optional configuration correctly.
+
+
+$ROSE_APP_OPT_CONF_KEYS
+-----------------------
+
+We could also specify the optional configurations to use by passing in an
+environment variable ``$ROSE_APP_OPT_CONF_KEYS``, a space delimited list
+of optional configuration names to use.
+
+Try using it by replacing the order_icecream task configuration with:
+
+.. code-block:: cylc
+
+   [[order_icecream]]
+       script = rose task-run
+       [[[environment]]]
+           ROSE_APP_OPT_CONF_KEYS = chocofudge
+
+If you re-run the suite, this should do the same job as the ``--opt-conf-key``
+method we tried before.
+
+
+Default Optional Configurations
+-------------------------------
+
+Optional configurations can be switched on by default by specifying their
+names via a top-level opts setting in the configuration file.
+
+Put the following line at the top of the order_icecream ``rose-app.conf`` file:
```

Comment:
'``app/order_icecream/rose-app.conf`` file' to get 'order_icecream' in monospace font.